### PR TITLE
Expose new prometheus metric discarded_samples_total

### LINKED ISSUE #1203: expose discarded samples metrics via prometheus /metrics endpoint
We use prometheus to scrap active_series counts on a per org basis so we can provide our customers with a unified dashboard that shows usage for both graphite (metrictank) and prometheus(cortex) hosted metrics services.

In addition to the current metrics, we also need to expose "discarded samples" numbers.
These metrics will be similar to the existing metrics we collect:
https://github.com/grafana/metrictank/blob/master/input/input.go#L47-L51
https://github.com/grafana/metrictank/blob/master/mdata/init.go#L31-L39

But will be a counter metric "discarded_samples_total" with labels for OrgID and discarded reason.
eg
```
var DiscardedSamples = prometheus.NewCounterVec(
	prometheus.CounterOpts{
		Namespace: "metrictank",
		Name: "discarded_samples_total",
		Help: "The total number of samples that were discarded.",
	},
	[]string{"reason", "org"},
)
```

the reason will be something like

- **out-of-order** (currently called `tank.metrics_too_old`)
- **new-value-for-timestamp** (currently called `tank.metrics_too_old` )
- **received-too-late** (currently called `tank.add_to_closed_chunk`)
- **invalid-timestamp** (currently called `input.%s.metricdata.invalid` or `input.%s.metricpoint.invalid`)
- **invalid-interval** (currently called `input.%s.metricdata.invalid`)
- **invalid-orgID** (currently called `input.%s.metricdata.invalid`)
- **invalid-name** (currently called `input.%s.metricdata.invalid`)
- **invalid-mtype** (currently called `input.%s.metricdata.invalid`)
- **invalid-tag-format** (currently called `input.%s.metricdata.invalid`)
- **unknown-point-id** (currently called `input.%s.metricpoint.unknown`)

### PR DESCRIPTION
**Pull request by fkaleo**
The metric takes a reason with values among:
- sample-out-of-order
- received-too-late
- invalid-timestamp
- invalid-interval
- invalid-orgID
- invalid-name
- invalid-mtype
- invalid-tag-format
- unknown-point-id

Fixes #1203

### CODE REVIEW

**Comment by fkaleo** [comment]
Depends on merging https://github.com/raintank/schema/pull/25

File: mdata/aggmetric.go [hunk 1]
```diff
@@ -409,14 +410,23 @@ func (a *AggMetric) Add(ts uint32, val float64) {
...
+		if err == nil {
+			if len(res) == 0 {
+				a.lastWrite = uint32(time.Now().Unix())
+			} else {
+				for _, p := range res {
+					a.add(p.Ts, p.Val)
+				}
+			}
+		} else {
+			var reason string
```
Comment (woodsaj) [inline mdata/aggmetric.go]: This needs to be initialised to something, eg "unknown".

Comment (fkaleo) [reply mdata/aggmetric.go]: fixed

File: mdata/aggmetric.go [hunk 2]
```diff
@@ -409,14 +410,25 @@ func (a *AggMetric) Add(ts uint32, val float64) {
...
+			switch err {
+			case errMetricsTooOld:
+				reason = sampleOutOfOrder
+			default:
+				reason = "unknown"
+			}
```
Comment (woodsaj) [inline mdata/aggmetric.go]: we should move the `metricsTooOld.Inc()` from ReorderBuffer.Add() to here as well.

Comment (fkaleo) [reply mdata/aggmetric.go]: Done

**Comment by fkaleo** [comment]
Manual testing can be performed using:
- the dashboard "Fakemetrics - discarded samples" included in this PR that works adequately with docker stack 'docker-dev-custom-cfg-kafka'
- `fakemetrics discarded` available with https://github.com/raintank/fakemetrics/pull/17

File: mdata/init.go [hunk 3]
```diff
@@ -15,6 +15,13 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+const (
+	discardReasonLabel = "reason"
```
Comment (Dieterbe) [inline mdata/init.go]: seems weird to declare this as a global since it only has 1 use, and for org we don't do this.

Comment (fkaleo) [reply mdata/init.go]: indeed! fixed.

File: input/input.go [hunk 4]
```diff
@@ -35,6 +36,17 @@ type DefaultHandler struct {
 	metricIndex idx.MetricIndex
 }
 
+// Possible reason labels for Prometheus metric discarded_samples_total
```
[the remark this answers is not in the document]
Comment (Dieterbe) [reply input/input.go]: no this is fine

**Comment by Dieterbe** [comment]
Note, this PR does not include `new-value-for-timestamp`. @fkaleo said he'll do that in a followup PR.

**Pull request merged**